In [214]:
import sympy
import numpy as np
import scipy.stats as ss
from scipy import stats
import matplotlib.pyplot as plt
import statsmodels
import statistics
%matplotlib inline
SEED = 1311

### Задача1

##### Пусть дана выборка x1, . . . ,xn из равномерного распределения на отрезке[a;θ]. Найдём оценку для неизвестного параметра θ методом моментов и методом максимального правдоподобия. Исследуем полученные оценки на несмещенность и состоятельность.Реализуем эту задачу:
1. сгенерируем выборку из равномерного распределения на [a;θ] размера n= 10, 100, 1000, 10 000;
2. найдём значения полученных оценок;
3. выведим отклонения полученных оценок от параметра θ. Что происходит с ростом n?

In [164]:
a = 28
b = a+a/2
size = [10,100,1000,10000][3]
uniform = np.random.uniform(low = a,high = b,size=size)


Используем метод моментов

In [165]:
print("Истинное значение параметра: ", b)
#Так как у нас один параметр, то рассмотрим момент первого порядка,то есть математическое ожидание EX.
#EX = (28+z)/2 , откуда мы и получаем уравнение на неизвестный параметр. Вместо теоретического момента EX подставим эмпирический момент - выборочное среднее x.
a_moment = sum(uniform)#Суммируем значения выборки
x = (1/size)*a_moment #Находим выборочное среднее
z_moment = (x*2)-28
print("Значение метода моментов: ",z_moment)

Истинное значение параметра:  42.0
Значение метода моментов:  42.11783273764364


Значение метода моментов размера выборки 10:     43.330884393990615
Значение метода моментов размера выборки 100:    42.6379421893577
Значение метода моментов размера выборки 1000:   41.942759551358634
Значение метода моментов размера выборки 10000:  42.11783273764364

Видим,что последнее значение максимально приближено к истинному значению. То есть ясно, что при увеличении выборки наше значение стремится к истинному, следовательно наша оценка состоятельна.

Метод максимального правдоподобия

In [166]:
print("Истинное значение параметра: ", b)
a_pravda = a+ss.uniform.fit(uniform)[0]/2
print("Значение метода максимального правдоподобия: ",a_pravda)

Истинное значение параметра:  42.0
Значение метода максимального правдоподобия:  42.000163433231435


Значение метода максимального правдоподобия размера выборки 10:    42.05821207387403
Значение метода максимального правдоподобия размера выборки 100:   42.041264934052876
Значение метода максимального правдоподобия размера выборки 1000:  42.000691281406205
Значение метода максимального правдоподобия размера выборки 10000: 42.000163433231435

Оценки, полученные методом максимального прадоподобия, будут состоятельными, асимптотически эффективными и асимптотически нормальными

In [167]:
print("Смещение: ", sum(uniform)/size-b)#смещение – это разность между математическим ожиданием оценки и истинным значением оцениваемого параметра, а несмещенная оценка – это оценка, имеющая нулевое смещение

Смещение:  -6.941083631178181


Видим, что оценка смещённая.

In [168]:
print("Отклонение метода моментов: ",{(b-z_moment) * (1 if b-z_moment>0 else -1)})
print("Отклонение метода максимального правдоподобия: ",{(b-a_pravda) * (1 if b-a_pravda>0 else -1)})

Отклонение метода моментов:  {0.11783273764363855}
Отклонение метода максимального правдоподобия:  {0.00016343323143530597}


Отклонение метода моментов при размере выборки 10:    1.3308843939906154
Отклонение метода моментов при размере выборки 100:   0.6379421893577017
Отклонение метода моментов при размере выборки 1000:  0.057240448641366015
Отклонение метода моментов при размере выборки 10000: 0.11783273764363855
Отклонение метода максимального правдоподобия при размере выборки 10:    0.05821207387403149
Отклонение метода максимального правдоподобия при размере выборки 100:   0.0412649340528759
Отклонение метода максимального правдоподобия при размере выборки 1000:  0.0006912814062047801
Отклонение метода максимального правдоподобия при размере выборки 10000: 0.00016343323143530597

Видно, что с увеличением выборки отклонение уменьшается, что логично, ведь чем больше мы проводим экспериментов,тем меньше погрешность.

### Задача2

##### Пусть дана реализация выборки x1, . . . ,xn из равномерного распределения на отрезке [θ; θ+ 1]. Найдите оценку для неизвестного параметра θ методом моментов и методом максимального правдоподобия.

Используем метод моментов

In [180]:
c = 28
d = c+1
size = [10,100,1000,10000][0]
uniform = np.random.uniform(low = c,high = d,size=size)


In [181]:
print("Истинное значение параметра: ", c)
#Так как у нас один параметр, то рассмотрим момент первого порядка,то есть математическое ожидание EX.
#EX = (28+z)/2 , откуда мы и получаем уравнение на неизвестный параметр. Вместо теоретического момента EX подставим эмпирический момент - выборочное среднее x
c_moment = sum(uniform)#Суммируем значения выборки
x = (1/size)*c_moment #Находим выборочное среднее
d_moment = (x*2)-28
print("Значение метода моментов: ",d_moment)

Истинное значение параметра:  28
Значение метода моментов:  29.025900198902406


In [182]:
print("Истинное значение параметра: ", c)
c_pravda = ss.uniform.fit(uniform)[0]+1
print("Значение метода максимального правдоподобия: ",c_pravda)

Истинное значение параметра:  28
Значение метода максимального правдоподобия:  29.079755269269928


In [185]:
print("Отклонение метода моментов: ",{(d-d_moment) * (1 if d-d_moment>0 else -1)})
print("Отклонение метода максимального правдоподобия: ",{(d-c_pravda) * (1 if d-c_pravda>0 else -1)})

Отклонение метода моментов:  {0.02590019890240569}
Отклонение метода максимального правдоподобия:  {0.07975526926992771}


### Задача 3

##### Пусть дана реализация выборки x1, . . . ,xn из нормального распределения N(θ1,θ2). Найдём оценки для неизвестных параметров θ1,θ2 методом моментов и методом максимального правдоподобия. Являются ли оценки θ1 несмещенными и состоятельными? Являются ли оценки θ2 несмещенными? Сделаем предположение о состоятельности оценок на θ2. Реализуем эту задачу:
1. сгенерируем выборку из нормального распределения N(θ1,θ2) размера n= 10, 100, 1000, 10 000;
2. найдём значения полученных оценок;
3. выведим отклонения полученных оценок от параметровθ1,θ2. Что происходит с ростом n?

In [284]:
m = 28
n = 1+m/2
size = [10,100,1000,10000][3]
norm = ss.norm.rvs(loc = m,scale = n,size=size)


Метод максимального правдоподобия

In [285]:
def L(args): return - \
    np.prod([ss.norm.pdf(xi, args[0], args[1]) for xi in norm])

#берём логарифм от правдоподобия


def logL(args): return - \
    np.sum([ss.norm.logpdf(xi, args[0], args[1]) for xi in norm])


In [286]:
from scipy.optimize import minimize
x0 = [0, 1]
# (-inf, inf) для первого параметра (мат.ож), (0, inf)  для второго (дисперсии)
bounds = [(None, None), (0, None)]
res = minimize(L, x0, bounds=bounds)
logres = minimize(logL, x0, bounds=bounds)

print(res.x)
print(logres.x)


[0. 1.]
[27.89785529 14.96634442]


Значение метода максимального правдоподобия размера выборки 10:    28.35764723 14.91437811
Значение метода максимального правдоподобия размера выборки 100:   27.2705931  15.79154028
Значение метода максимального правдоподобия размера выборки 1000:  28.32139186 15.28843083
Значение метода максимального правдоподобия размера выборки 10000: 27.89785529 14.96634442

Оценки, полученные методом максимального прадоподобия, будут состоятельными, асимптотически эффективными и асимптотически нормальными

Метод моментов

In [287]:
#Для отыскания двух неизвестных параметров необходимо иметь два уравнения:
#приравняем начальный теоретический момент первого порядка и центральный теоретический момент второго порядка соответствующим эмпирическим моментам
#v1= M(Х),М1=х_vub,µ2 = D(X),m2 =D(B), получаем систему M(Х)= x_vub
#                                                       D(X) = D(B)
#Так как математическое ожидание и дисперсия нормального распределения известны в ввиде начальных параметрых, тогда получаем систему
#M(Х)=a= x_vub
#D(X)=σ**2 = D(B)
from cmath import sqrt
a = 1/size*sum(norm)#выборочное среднее
r = stats.moment(norm, moment=2, axis=0)#находим 2 момент, который соответствует дисперсии
σ = sqrt(r)
print(a)
print(σ)


27.897880014536245
(14.96631990081778+0j)


Значение метода моментов размера выборки 10:    28.357692179600292 14.91433647250344
Значение метода моментов размера выборки 100:   27.27046454327667  15.791573864011605
Значение метода моментов размера выборки 1000:  28.321428519400712 15.288421058907177
Значение метода моментов размера выборки 10000: 27.897880014536245 14.96631990081778

Видим,что значения стремятся к истинному параметру, значит оценки состоятельны.

In [288]:
print("Смещение1 = ", sum(norm)/size-m)
print("Смещение2 = ", sum(norm)/size-n)


Смещение1 =  -0.10211998546375511
Смещение2 =  12.897880014536245


Видим, что для 1 параметра оценка - несмещённая, а для 2 параметра - смещённая.